src/workload.rs

```diff
@@ -144,32 +174,108 @@ impl Workload for DoubleInsertion {
                 format!("1*return(\"{}\")", injection),
             )
             .await?;
-            conn.query_drop("BEGIN OPTIMISTIC")?;
+            send!(conn, "BEGIN OPTIMISTIC")?;
             let row = table.new_row();
             let insertion_1 = format!("INSERT INTO {} VALUES ({})", table.name, row.to_string());
             let insertion_2 = format!(
                 "INSERT INTO {} VALUES ({})",
                 table.name,
                 row.next().to_string()
             );
-            println!("execute: {}", insertion_1);
-            let res = conn
-                .query_drop(insertion_1)
-                .and_then(|_| {
-                    println!("execute: {}", insertion_2);
-                    conn.query_drop(insertion_2)
-                })
-                .and_then(|_| {
-                    println!("execute: commit");
-                    conn.query_drop("commit")
-                });
+
+            let res = async {
+                send!(conn, query(insertion_1.as_str()))?;
+                send!(conn, query(insertion_2.as_str()))?;
+                Ok(())
+            }
+            .await;
+
+            if res.is_err() {
+                send!(conn, "ROLLBACK")?;
+            }
+            info!(log, "{:?}", res);
+
+            collect_result(res, results, &table, injection, pool.clone()).await;
+            disable_failpoint(
+                client,
+                "github.com/pingcap/tidb/table/tables/corruptMutations",
+            )
+            .await?;
+        }
+
+        Ok(())
+    }
+}
+
+// 2 txns. The first writes corrupted data and the second reads it. Check if Assertion can detect it.
+struct T2;
+#[async_trait]
+impl Workload for T2 {
+    async fn execute(
+        &self,
+        log: Logger,
+        config: &Config,
+        table: Table,
+        client: &reqwest::Client,
+        pool: Arc<Pool<MySql>>,
+        results: &mut HashMap<(Table, String, String), Effectiveness>,
+    ) -> Result<()> {
+        let mut conn = pool.acquire().await?;
+        enable_featuers(&mut conn, config).await?;
+        let drop_statement = table.drop_statement();
+        let create_statement = table.create_statement();
+        let start = Instant::now();
+        send!(conn, query(drop_statement.as_str())).expect("don't let drop statement fail");
+        send!(conn, query(create_statement.as_str())).expect("don't let create statement fail");
+        let duration = start.elapsed();
+        CREATE_TABLE_DURAION_MS.fetch_add(duration.as_millis() as u64, Ordering::SeqCst);
+        info!(log, "{}", create_statement);
+
+        for injection in AVAILABLE_INJECTIONS {
+            info!(log, "{} ready to go!", injection);
+
+            // NOTE: "1*" here, otherwise an index mutation is missing for each row insertion, thus cannot be detected.
+            enable_failpoint(
```

Comment:
Actually, I think you can consider inject it on either the `insert` or the `update`. I think assertion should possibly take effect on both cases.

Reply:
OK let's make it another workload.